### PR TITLE
Added landline error handler

### PR DESCRIPTION
**Pull request by johanna-skylight**
# BACKEND PULL REQUEST

## Related Issue

Fixes [#3723](https://github.com/CDCgov/prime-simplereport/issues/3723)

## Changes Proposed

- Implement an error handler for events with code Twilio 30006 (landline number). 
- Add a conditional in the Twilio event controller (AKA SmsCallbackController) that checks for error codes and when 30006 error is detected it calls the landline error handler.
- When a phone number is detected as landline we looks for those records in the phone number table and retrieve them is they are mistakenly marked as mobile. 
- We then update the type for all the retrieved records and push the landline update.

## Additional Information
Text coverage was also added as part of this PR.

## Testing
Make Twilio to send an error event to the SimpleReport callback with landline error code 30006 and check that the phone number record type has properly been updated to LANDLINE.

## Checklist for Primary Reviewer

- [ ] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke tested
- [X] Changes comply with the SimpleReport Style Guide

### CODE REVIEW

**Comment by johanna-skylight** [comment]
@zdeveloper  I saw you wrote some unit test using mockito so it would be awesome if you could check on the testing :D

**Review by emmastephenson — DISMISSED** [review]
This looks great Johanna, thank you!

Were you able to verify that after attempting to send the text to a landline, the patient profile also silently updated the phone type as landline? (No reason to think it wouldn't, I just want to double check!)

**Comment by johanna-skylight** [comment]
> This looks great Johanna, thank you!
> 
> Were you able to verify that after attempting to send the text to a landline, the patient profile also silently updated the phone type as landline? (No reason to think it wouldn't, I just want to double check!)

In general it will not matter how the text message was sent to the person because the way that twilio reports these errors to SimpleReport is async. When SimpleReport asks Twilio to send a message it just "queues" it in Twilio messaging service and only receives an OK(200) acknowledging that the messages was received by Twilio and will be sent, but at this point the message is not actually sent to the end user.  

Afterwards Twilio will process the requests in its messaging service and will notify SimpleReport of errors, warnings and messages being processed by communicating to the /api/pxp/callback endpoint. Is in this callback that the new logic has been implemented. At this point we detect the error code 3006 and proceed to scan the phone number table and adjust the phone type for the landline number that triggered the Twilio error. 

Not sure if that answered the question. I would need more details on this patient profile scenario. Is it a specific table? or a specific flow?.

**Comment by emmastephenson** [comment]
> > This looks great Johanna, thank you!
> > Were you able to verify that after attempting to send the text to a landline, the patient profile also silently updated the phone type as landline? (No reason to think it wouldn't, I just want to double check!)
> 
> In general it will not matter how the text message was sent to the person because the way that twilio reports these errors to SimpleReport is async. When SimpleReport asks Twilio to send a message it just "queues" it in Twilio messaging service and only receives an OK(200) acknowledging that the messages was received by Twilio and will be sent, but at this point the message is not actually sent to the end user.
> 
> Afterwards Twilio will process the requests in its messaging service and will notify SimpleReport of errors, warnings and messages being processed by communicating to the /api/pxp/callback endpoint. Is in this callback that the new logic has been implemented. At this point we detect the error code 3006 and proceed to scan the phone number table and adjust the phone type for the landline number that triggered the Twilio error.
> 
> Not sure if that answered the question. I would need more details on this patient profile scenario. Is it a specific table? or a specific flow?.

Ah I'm sorry, I should've been more specific! The patient profile is the same as the add/edit patient page. It's where end users can change phone numbers, phone types, and preferences. 
Screen Shot 2022-05-18 at 8 32 17 AM

By saving the new type in the database through the async Twilio flow, this page should automatically populate with landline the next time someone opens up the patient profile after the async change has been made. There's no reason to believe that it wouldn't do this with the code you have written, I just like to double-check.

**Comment by johanna-skylight** [comment]
> 

Yeah, next time that phone number is retrieve it should display as Landline for any person

File: backend/src/main/java/gov/cdc/usds/simplereport/service/sms/TextMessageStatusService.java [hunk 1]
```diff
@@ -74,4 +84,13 @@ private List<String> getQueryStringKeys(String queryString) {
           .collect(Collectors.toList());
     }
   }
+
+  @Transactional
+  public void handleLandlineError(String landlineNumber) {
```
Comment (nickclyde) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/sms/TextMessageStatusService.java:89]: Love this method, clean and simple.

File: backend/src/test/java/gov/cdc/usds/simplereport/service/sms/TextMessageStatusServiceTest.java [hunk 2]
```diff
@@ -53,4 +83,51 @@ void saveTextMessageStatus_success() {
...
+        .thenReturn(testNumbers);
+
+    ArgumentCaptor<List> phoneNumbersCaptor = ArgumentCaptor.forClass(List.class);
+    textMessageStatusService.handleLandlineError(toParam);
+
+    verify(_phoneRepo).findAllByNumberAndType(number, PhoneType.MOBILE);
+    verify(_phoneRepo).saveAll(phoneNumbersCaptor.capture());
+    List<PhoneNumber> modifiedNumbers = phoneNumbersCaptor.getValue();
+    assertEquals(1, modifiedNumbers.size());
+    assertEquals(PhoneType.LANDLINE, modifiedNumbers.get(0).getType());
```
Comment (nickclyde) [inline backend/src/test/java/gov/cdc/usds/simplereport/service/sms/TextMessageStatusServiceTest.java:131]: Great tests!

**Review by nickclyde — DISMISSED** [review]
Awesome work Johanna!

File: backend/src/main/java/gov/cdc/usds/simplereport/service/sms/TextMessageStatusService.java [hunk 3]
```diff
@@ -74,4 +86,44 @@ private List<String> getQueryStringKeys(String queryString) {
           .collect(Collectors.toList());
     }
   }
+
+  private String getNumberByMessageId(String messageId, String twilioNumber) {
```
Comment (johanna-skylight) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/sms/TextMessageStatusService.java:90]: @emmastephenson after our Twilio troubleshooting session I realized that the number sent in the status request is masked (as you pointed out). So I had to add a method to retrieve the phone number using the messageId like once we discussed. Please let me know if any questions.

File: backend/src/test/java/gov/cdc/usds/simplereport/service/sms/TextMessageStatusServiceTest.java (hunk 2 above)
Comment (emmastephenson) [reply backend/src/test/java/gov/cdc/usds/simplereport/service/sms/TextMessageStatusServiceTest.java:150]: +1, these are perfect!

**Review by emmastephenson — APPROVED** [review]
Looks great, thank you Johanna!

**Pull request merged**